airflow/providers/microsoft/azure/hooks/azure_batch.py

```diff
@@ -0,0 +1,348 @@
+
+        """
+        try:
+            self.log.info("Attempting to create a pool: %s", pool.id)
+            self.connection.pool.add(pool)
+            self.log.info("Created pool: %s", pool.id)
+        except batch_models.BatchErrorException as e:
+            if e.error.code != "PoolExists":
+                raise
+            else:
+                self.log.info("Pool %s already exists", pool.id)
+
+    def _get_latest_verified_image_vm_and_sku(self, publisher, offer, sku_starts_with):
+        """
+        Get latest verified image vm and sku
+
+        :param publisher: The publisher of the Azure Virtual Machines Marketplace Image.
+            For example, Canonical or MicrosoftWindowsServer.
+        :type publisher: str
+        :param offer: The offer type of the Azure Virtual Machines Marketplace Image.
+            For example, UbuntuServer or WindowsServer.
+        :type offer: str
+        :param sku_starts_with: The start name of the sku to search
+        :type sku_starts_with: str
+        """
+
+        options = batch_models.AccountListSupportedImagesOptions(
+            filter="verificationType eq 'verified'")
+        images = self.connection.account.list_supported_images(
+            account_list_supported_images_options=options)
+        # pick the latest supported sku
+        skus_to_use = [
+            (image.node_agent_sku_id, image.image_reference) for image in images
+            if image.image_reference.publisher.lower() == publisher.lower() and
+            image.image_reference.offer.lower() == offer.lower() and
+            image.image_reference.sku.startswith(sku_starts_with)
+        ]
+
+        # pick first
+        agent_sku_id, image_ref_to_use = skus_to_use[0]
+        return agent_sku_id, image_ref_to_use
+
+    def wait_for_all_node_state(self, pool_id, node_state):
+        """
+        Wait for all nodes in a pool to reach given states
+
+        :param pool_id: A string that identifies the pool
+        :type pool_id: str
+        :param node_state: A set of batch_models.ComputeNodeState
+        :type node_state: set
+        """
+        self.log.info('waiting for all nodes in pool %s to reach one of: %s',
+                      pool_id, node_state)
+        i = 0
+        while True:
+            # refresh pool to ensure that there is no resize error
+            pool = self.connection.pool.get(pool_id)
+            if pool.resize_errors is not None:
+                resize_errors = "\n".join([repr(e) for e in pool.resize_errors])
+                raise RuntimeError(
+                    'resize error encountered for pool {}:\n{}'.format(
+                        pool.id, resize_errors))
+            nodes = list(self.connection.compute_node.list(pool.id))
+            if (len(nodes) >= pool.target_dedicated_nodes and
+                    all(node.state in node_state for node in nodes)):
+                return nodes
+            i += 1
+            if i % 3 == 0:
```

Comment:
Why the magic constant :)?

Reply:
Mistake. I wanted to raise exception after checking the states 3 times. I can see I didn't raise it. I'll do the correction. Thanks for the review